src/features/mcda/atoms/mcdaCalculation.ts

```diff
@@ -0,0 +1,223 @@
+  sentiment: [string, string],
+  coefficient: number,
+) => {
+  const layerNormalizedFormula = [
+    '*',
+    coefficient,
+    ['/', ['-', ['/', ['get', num], ['get', den]], min], max - min],
+  ];
+
+  if (isEqual(sentiment, sentimentDefault)) {
+    return layerNormalizedFormula;
+  } else if (isEqual(sentiment, sentimentReversed)) {
+    return ['-', 1, layerNormalizedFormula];
+  }
+};
+
+const linearNormalization = (json: JsonMCDA) => {
+  const layersCount = json.axes.length;
+  if (layersCount === 1) {
+    return layerNormalized(
+      json.axes[0],
+      json.ranges[0],
+      json.sentiments[0],
+      json.coefficients[0],
+    );
+  } else if (layersCount > 1) {
+    return [
+      '/',
+      [
+        '+',
+        ...json.axes.map((_, i) =>
+          layerNormalized(
+            json.axes[i],
+            json.ranges[i],
+            json.sentiments[i],
+            json.coefficients[i],
+          ),
+        ),
+      ],
+      sum(json.coefficients),
+    ];
+  }
+};
+
+export const mcdaCalculationAtom = createAtom(
+  {
+    calcMCDA: (json: JsonMCDA) => json,
+    enableMCDALayer: (layerId: string) => layerId,
+    disableMCDALayer: () => null,
+  },
+  ({ onAction, schedule, getUnlistedState, create }) => {
+    onAction('calcMCDA', (json) => {
+      const id = json.id || 'MCDA_layer';
+      const { good = DEFAULT_GREEN, bad = DEFAULT_RED } = json.colors;
+      const layerStyle: BivariateLayerStyle = {
+        id,
+        type: 'fill',
+        layout: {},
+        filter: filterSetup(json.axes),
+        paint: {
+          'fill-color': [
+            'let',
+            'mcdaResult',
+            ['to-number', linearNormalization(json), -1], // falsy values become -1
+            [
+              'case',
+              ['all', ['>=', ['var', 'mcdaResult'], 0], ['<=', ['var', 'mcdaResult'], 1]],
+              ['interpolate', ['linear'], ['var', 'mcdaResult'], 0, bad, 1, good],
+              'transparent', // all values outside of range [0,1] will be painted as transparent
+            ],
+          ],
+          'fill-opacity': 1,
+          'fill-antialias': false,
+        },
+
+        source: {
+          type: 'vector',
+          tiles: [
+            `${adaptTileUrl(
+              appConfig.bivariateTilesRelativeUrl,
+            )}{z}/{x}/{y}.mvt?indicatorsClass=${appConfig.bivariateTilesIndicatorsClass}`,
+          ],
+          maxzoom: 8,
+          minzoom: 0,
+        },
+        'source-layer': 'stats',
+      };
+
+      if (layerStyle) {
+        const layerSource = layerStyle.source;
+        const source = layerSource
+          ? {
+              id,
+              maxZoom: layerSource.maxzoom,
+              minZoom: layerSource.minzoom,
+              source: {
+                type: layerSource.type,
+                urls: layerSource.tiles,
+                tileSize: 512,
+                apiKey: '',
+              },
+            }
+          : undefined;
```

Comment:
1. you had `layerStyle` declared as object on a block above so that `if` is pointless
2. why do you conditionally take source from layer style considering you've just manually declared it befor?
3. why fallback condition isn't just declared as layer source initially?

Reply:
good points, didn't pay attention as it's was just a copy of code from src/features/bivariate_manager/atoms/bivariateMatrixSelection.ts
will fix